runtime/runtime_test.go

```diff
@@ -373,51 +379,153 @@ func TestRuntimeImport(t *testing.T) {
+			checkCount += 1
+		},
+		setProgram: func(location Location, program *interpreter.Program) error {
+			programs[location.ID()] = program
+			return nil
+		},
+		getProgram: func(location Location) (*interpreter.Program, error) {
+			return programs[location.ID()], nil
+		},
 	}
 
-	nextTransactionLocation := newTransactionLocationGenerator()
+	const transactionCount = 10
+	for i := 0; i < transactionCount; i++ {
 
-	value, err := runtime.ExecuteScript(
-		Script{
-			Source: script,
+		nextTransactionLocation := newTransactionLocationGenerator()
+
+		value, err := runtime.ExecuteScript(
+			Script{
+				Source: script,
+			},
+			Context{
+				Interface: runtimeInterface,
+				Location:  nextTransactionLocation(),
+			},
+		)
+		require.NoError(t, err)
+
+		assert.Equal(t, cadence.NewInt(42), value)
+	}
+	require.Equal(t, transactionCount+1, checkCount)
+}
+
+func TestRuntimeConcurrentImport(t *testing.T) {
+
+	t.Parallel()
+
+	runtime := NewInterpreterRuntime()
+
+	importedScript := []byte(`
+      pub fun answer(): Int {
+          return 42
+      }
+    `)
+
+	script := []byte(`
+      import "imported"
+
+      pub fun main(): Int {
+          let answer = answer()
+          if answer != 42 {
+            panic("?!")
+          }
+          return answer
+        }
+    `)
+
+	var checkCount uint64
+	var programsLock sync.RWMutex
+	programs := map[common.LocationID]*interpreter.Program{}
+
+	runtimeInterface := &testRuntimeInterface{
+		getCode: func(location Location) (bytes []byte, err error) {
+			switch location {
+			case common.StringLocation("imported"):
+				return importedScript, nil
+			default:
+				return nil, fmt.Errorf("unknown import location: %s", location)
+			}
 		},
-		Context{
-			Interface: runtimeInterface,
-			Location:  nextTransactionLocation(),
+		programChecked: func(location common.Location, duration time.Duration) {
+			atomic.AddUint64(&checkCount, 1)
 		},
-	)
-	require.NoError(t, err)
+		setProgram: func(location Location, program *interpreter.Program) error {
+			programsLock.Lock()
+			defer programsLock.Unlock()
+
+			programs[location.ID()] = program
+
+			return nil
+		},
+		getProgram: func(location Location) (*interpreter.Program, error) {
+			programsLock.RLock()
+			defer programsLock.RUnlock()
+
+			program := programs[location.ID()]
+
+			return program, nil
+		},
+	}
+
+	var wg sync.WaitGroup
+	const concurrency uint64 = 10
+	for i := uint64(0); i < concurrency; i++ {
+		nextTransactionLocation := newTransactionLocationGenerator()
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			value, err := runtime.ExecuteScript(
+				Script{
+					Source: script,
+				},
+				Context{
+					Interface: runtimeInterface,
+					Location:  nextTransactionLocation(),
+				},
+			)
+			require.NoError(t, err)
 
-	assert.Equal(t, cadence.NewInt(42), value)
+			assert.Equal(t, cadence.NewInt(42), value)
+		}()
+	}
+	wg.Wait()
+
+	//TODO:
+	//require.Equal(t, concurrency+1, checkCount)
```

Comment:
What's the blocker here?

Reply:
Ideally we would expect the imported program only be checked once (`concurrency` transactions + 1 for the imported program), however, currently the imported program gets re-checked if it is currently being checked. 

I think this can be optimized by synchronizing the checking of a program using `sync`, but I felt like that was a future optimization that could be done in another PR.

I've added this explanation as a comment in 762985cf5693181aa6cfdde9461a67e0c09e869b